repo.go

```diff
@@ -688,71 +983,177 @@ func (r *Repo) AddTargetsWithDigest(digest string, digestAlg string, length int6
+	// corresponding targets manifests.
+	targetsMetaToWrite := map[string]*targetsMetaWithSigners{}
+
 	if err := r.local.WalkStagedTargets(normalizedPaths, func(path string, target io.Reader) (err error) {
-		meta, err := util.GenerateTargetFileMeta(target, r.hashAlgorithms...)
+		targetsMeta, delegation, err := r.targetDelegationForPath(path, preferredRole)
+		if err != nil {
+			return err
+		}
+		// This is the targets role that needs to sign the target file.
+		targetsRoleName := delegation.Delegatee.Name
+
+		signers, err := r.getSignersInDB(targetsRoleName, delegation.DB)
+		if err != nil {
+			return err
+		}
+
+		// FIXME: get a complete set of all possible signers, not just the ones
+		// along the delegation path used to reach the preferredRole.
+
+		// The delegations determine the keys that need to be used to sign each
+		// target file. If a particular target role is reachable in the delegation
+		// graph through multiple paths, it may be the case that the targets
+		// manifest needs signatures from multiple sets of keys. Therefore, we
+		// merge the keys specified by the current delegation with keys specified
+		// by previously seen delegations.
+		if prevMetaWithSigners, ok := targetsMetaToWrite[targetsRoleName]; ok {
+			// Start with meta staged in targetsMetaToWrite instead of the committed meta.
+			targetsMeta = prevMetaWithSigners.meta
+
+			mergedKeyIDs := map[string]struct{}{}
+			for _, signer := range signers {
+				for _, keyID := range signer.PublicData().IDs() {
+					mergedKeyIDs[keyID] = struct{}{}
+				}
+			}
+
+			// Merge prevMetaWithSigners.signers with signers.
+			for _, prevSigner := range prevMetaWithSigners.signers {
+			keyIDLoop:
+				for _, keyID := range prevSigner.PublicData().IDs() {
+					if _, seen := mergedKeyIDs[keyID]; !seen {
+						signers = append(signers, prevSigner)
+						mergedKeyIDs[keyID] = struct{}{}
+						break keyIDLoop
+					}
+				}
+			}
+		}
+
+		fileMeta, err := util.GenerateTargetFileMeta(target, r.hashAlgorithms...)
 		if err != nil {
 			return err
 		}
-		path = util.NormalizeTarget(path)
 
 		// if we have custom metadata, set it, otherwise maintain
 		// existing metadata if present
 		if len(custom) > 0 {
-			meta.Custom = &custom
-		} else if t, ok := t.Targets[path]; ok {
-			meta.Custom = t.Custom
+			fileMeta.Custom = &custom
+		} else if tf, ok := targetsMeta.Targets[path]; ok {
+			fileMeta.Custom = tf.Custom
 		}
 
 		// G2 -> we no longer desire any readers to ever observe non-prefix targets.
-		delete(t.Targets, "/"+path)
-		t.Targets[path] = meta
+		delete(targetsMeta.Targets, "/"+path)
+		targetsMeta.Targets[path] = fileMeta
+
+		targetsMetaToWrite[targetsRoleName] = &targetsMetaWithSigners{
+			meta:    targetsMeta,
+			signers: signers,
+		}
+
 		return nil
 	}); err != nil {
 		return err
 	}
-	return r.writeTargetWithExpires(t, expires)
-}
 
-func (r *Repo) writeTargetWithExpires(t *data.Targets, expires time.Time) error {
```

Comment:
While making `AddTargetsWithDigest` support delegated targets, I removed `writeTargetWithExpires` since it was only used in two places (`AddTargetsWithDigest` and `Init`). Versions numbers are explicitly bumped/set everywhere else in the code, and having the side effect of mutating the passed `data.Targets` is slightly confusing with that convention.